minimap overhaul

stonerl opened the pull request:
* buttons for the minimap are in one row
* minimap zoom2 limited to 600px
* minimap tries to scale up to 300px/400px in normal mode
* when map has a width of more than 300, zoom is disabled
* disable zoom if the minimap height gets greater than the game's
* check size when switching into window mode

fixes #528

stonerl committed:
overhaul minimap

* buttons for the minimap are in one row
* minimap zoom2 limited to 600px
* minimap tries to scale to 300px
* when map has a width of more than 300, zoom is disabled
* center minimap window on zoom

fixes #528

stonerl committed:
disable zoom if the minimap windows gets higher than the actual game

stonerl commented:
Some pictures:
![shot0001](https://user-images.githubusercontent.com/2091312/70726536-88c36880-1cfe-11ea-92c2-769f394d0495.png)
![shot0002](https://user-images.githubusercontent.com/2091312/70726538-88c36880-1cfe-11ea-82f5-5466a8a0efbc.png)
![shot0003](https://user-images.githubusercontent.com/2091312/70726541-895bff00-1cfe-11ea-93c7-8160a6312977.png)
![shot0004](https://user-images.githubusercontent.com/2091312/70726542-895bff00-1cfe-11ea-9291-c1644a48a594.png)
![shot0005](https://user-images.githubusercontent.com/2091312/70726544-895bff00-1cfe-11ea-8bb1-6238c1e56e90.png)

stonerl committed:
scale maps as much as possible

stonerl committed:
zoom scales as much as possible and not only 2 times

stonerl committed:
removed superfluous if

stonerl committed:
make sure minimap window isn't higher than the game itself

Notabilis reviewed (changes requested):
Overall testing worked good, nice to have a usable minimap on small maps.
Some observations:
- The "red box" that should display the current viewport is too small, no idea what it is showing. This is also visible in your demo-screenshots, e.g., the first one: The red box is a small rectangle in the middle of the player's territory but the viewport even shows the borders. Please fix this one, my further observations are "nice to have".
- The zoom button is still disabled after switching to fullscreen. Also, the map might be "lost" outside the screen when switching back to window mode, but this is a problem of all dialogs and most likely out of scope of this branch.
- The "is zoomed" state is stored and restored when re-opening the window, even when the resolution of the game window is now too small to contain the zoomed minimap.
- Just a thought: The minimap window might be too large for some players, in case they prefer to keep it always open in a corner. But we can deal with that if anyone complains.
- For small map, the minimap does look ugly/pixelated. But well, so be it. Personally, I like it, but no idea what younger players think of it. ^^
- An old bug, already in trunk: After switching to fullscreeen mode, the red box is only recalculated when the map is moved (even though the game window now displays more of the game map)

The code seems to be okay, though I haven't understood everything and was too lazy to dig deeper. But it is UI code and the result looks good, so I guess it is alright.
You have a lot of `?:` clauses in the code, I think all of them can be replaced by passing the condition directly as parameter to `scale_map()`. Or is there a reason that you still have them in the code?

stonerl committed:
remove if-clause where possible and pass zoom as an argument

stonerl committed:
fix viewport frame size

stonerl commented:
@Notabilis The viewport problem should be solved. All the other issues need to be fixed otherwhere. Since this is a general problem with window resizing, that happens to all ingame windows.

stonerl committed:
don't center minimpa on every opening

Notabilis reviewed (dismissed):
Thanks for the changes, looks good now.
Can you please add your explanations about the pixel sizes as a comment to the code?
Beside that, feel free to merge as soon as the automatic checks are fine.

stonerl committed:
resize minimap on resolution change, if requirred

fixes #528

stonerl committed:
resize minimap on opening if necessary

stonerl commented:
@Notabilis would you test again.

>     * The zoom button is still disabled after switching to fullscreen. Also, the map might be "lost" outside the screen when switching back to window mode, but this is a problem of all dialogs and most likely out of scope of this branch.
> 
fixed
>     * The "is zoomed" state is stored and restored when re-opening the window, even when the resolution of the game window is now too small to contain the zoomed minimap.
> 

fixed
>     * An old bug, already in trunk: After switching to fullscreeen mode, the red box is only recalculated when the map is moved (even though the game window now displays more of the game map)
> 
 Currently, no idea how to fix it.

Notabilis commented:
Nice, I didn't expect Widelands to offer such a notification. Its looking good!

A small, most likely old bug I noticed: When opening the window, the state of the buttons does not represent their logical state (i.e., they are not pressed down but the terrain is drawn anyway). This can be fixed by adding `update_button_permpressed();` in wui/minimap.cc after line 233, but I haven't checked for side effects. Also, I haven't checked why they become "unpressed" at all, so there might be a cleaner fix for it.

For the third bug you cited ("An old ..."), it seems to be enough to add `mainview_move();` after wui/interactive_base.cpp line 229 within the GraphicResolutionChanged notification handler.

stonerl committed:
addressed review comments

stonerl commented:
@Notabilis should be fixed now.

Notabilis reviewed (approved):
Thanks, works fine. I wasn't able to find any further bugs, so feel free to merge when the tests are done.
